### ISSUE TITLE
Disappointing performance

### ISSUE THREAD

**Issue by makuto** [Random06457/rpi4-gpu-bare-metal-examples#5]
I've been making some modifications in [my fork](https://macoy.me/code/macoy/rpi-gpu-bare-metal). I'm trying to get a reasonable frame rate, but am having troubles and was hoping you could provide me some insight as to why.

My target is 100,000 triangles rendered at 60 hz 1080p with the shaders you have hard-coded. I don't know if this is a reasonable target, but it is a practical one for the goal of hardware accelerated UI, 2D games, and low poly 3D games. Unfortunately, I'm only seeing about 3 hz with only 10,000 triangles on the 500x500 canvas. The [Broadcom VideoCore IV manual](https://docs.broadcom.com/doc/12358545) claims the following:
> The main specification points for a fully configured system at 250 MHz are:
> • 25M rendered triangles/s.
> • 1G pixels/s with single bilinear texturing, simple shading, 4x multisampling.

So I thought my 100k tris per frame (well under the 400k tris/frame at 60hz claimed) was reasonable, especially if this is now the VideoCore VI, which should be faster than the IV.

# My modifications
- Render 10,000 vertices so it's approaching a realistic workload.
- Re-use the binning and render lists rather than rebuild them every frame
- Put the binning and render execution into a loop.
- Update the vertex colors and re"upload" them (which is an unnecessary copy; I could build the colors straight in the buffer, but with this amount of memory I would hope it would not affect the frame rate that much), which is important because otherwise I won't be able to tell the frame has changed
- Render straight to the framebuffer. It appears blitting the render target to the framebuffer with the CPU is quite slow.
- My fork also has some changes to make the `aarch64-none-elf-` toolchain work so I don't need `gnu`.

# Issues I saw
- The `tile_alloc_size` needs to be grown according to how much attribute data there is. For 100,000 vertices I had to actually make it 2x the size of the vertex and color buffers; the relation does not appear to be 1:1. Any insight as to how this alloc size relates to the attributes size would be much appreciated!
- [This line](https://github.com/Random06457/rpi4-gpu-bare-metal-examples/blob/master/src/kernel/examples/gl_rainbow_triangle.cpp#L498C5-L498C5) and the corresponding `getRfCount` need to be updated to handle wrapping. Once the frame count wraps from 255 to 0, the program will get stuck waiting for a greater number than 255. I made the types `u8` so it is more clear this can happen.

It appears the allocator is broken. If I put the frame loop around the call to `rainbowTriangleGL()`, the allocator fails to allocate the CLEmitters on the second frame. I think there may be a problem with the linked list nodes in the heap allocator. I didn't look into it because I wanted to focus on the graphics stuff, so I just re-use the already allocated buffers instead.

# Ideas
I removed the `v3dInvalidateCaches()` in the render loop, but I'm still at 3 hz. Do you have any ideas for what might be causing the poor performance?

**Comment by Random06457** [comment]
Hi,
I didn't do much profiling so far since I was really just focused on getting a single frame to render.

However one thing that could greatly improve speed could be enabling the MMU to enable CPU caches which I don't think are enabled yet (?). This is most likely why blitting takes so much time.

In terms of algorithms etc I don't really remember, I'd have to investigate.

By the way, it seems like you are actively working on this, if you want more spontaneous/informal answers, feel free to add me on discord at `random6457`

**Comment by makuto** [comment]
I don't use Discord, but I do have Matrix: `@macoy:matrix.org`.
I will try it with the MMU, but I don't expect it to make a huge difference because I wouldn't expect the VPU to obey the CPU's memory rules (e.g. normal vs device). I'm going to try cutting the CPU part down as much as I can to see if it is the CPU or the VPU which is the bottleneck. I thought maybe the error I saw `VCD_IDLE` in `V3D_ERR_STAT` could be an issue (the VCD does  DMA reads), but I read in the VC4 manual that the VCD is automatically used for the vertex pipeline.

**Comment by makuto** [comment]
Can you explain the Z buffer stride? It is allocated like this:
`V3DBuffer z_buffer(width * height * 16, 4096);`
Then it is used here:
``` C++
indirect << StoreTileBufferGeneral(
						BUFFER_Z, V3D_TILING_UIF_XOR, false, V3D_DITHER_MODE_NONE,
						V3D_DECIMATE_MODE_SAMPLE_0, V3D_OUTPUT_IMAGE_FORMAT_D16,
						false, false, false, (height + 15) / (2 * 4), 0,
						z_buffer.paddr());
```
I'm scratching my head at this equation: `(height + 15) / (2 * 4)`. If it is `D16`, why isn't the stride the `renderWidth * sizeof(D16)`, or `renderWidth * 2` (2 bytes per depth value)?

**Comment by makuto** [comment]
Another question for you. I've been making a single-header version of v3d [here](https://macoy.me/code/macoy/rpi-system/src/branch/master/addon/v3d/v3d.h), which for now also needs [this header](https://macoy.me/code/macoy/rpi-bare-metal/src/branch/master/tools/v3d.h) (I will paste that in once it works). The sample usage is [here](https://macoy.me/code/macoy/rpi-system/src/branch/master/sample/001-v3d/kernel.cpp).

I've carefully combed through my command lists and they are nearly bit-for-bit identical to yours, except my vertex counts and addresses are different (as expected).

However, I'm getting stuck at the start of binning. The flush count never increases despite my command list having the flush operation at the end. Did you encounter anything similar when writing your version, and if so, how did you debug it? I tried checking the `V3D_CLE_CT0CS`, `V3D_ERR_STAT`, and `V3D_CLE_PCS`, but they all seem nominal.

**Comment by makuto** [comment]
I have managed to make progress by gradually taking your working example and introducing my suspect code until I've weeded out all the issues in my port.

**Comment by makuto** [comment]
I have good news! I got V3D working from the Circle setup. I stumbled with the MMU being on and my forgetting to flush the caches so the GPU could read my command lists. Once that was sorted I was able to render.

I think my test case was close to worst-case, where every triangle nearly overlaps the entire render target. It is able to render 100,000 triangles evenly spaced (non-overlapping) throughout a 750x750 pixel render target very quickly. I'm going to try 1080p next and then subsequently get a real 3D (untextured) scene in to see how it goes.
